rclcpp/src/rclcpp/context.cpp

```diff
@@ -341,20 +348,38 @@ Context::shutdown(const std::string & reason)
   return true;
 }
 
-rclcpp::Context::OnShutdownCallback
-Context::on_shutdown(OnShutdownCallback callback)
+rclcpp::OnShutdownCallbackHandle::SharedPtr
+Context::on_shutdown(OnShutdownCallbackHandle::OnShutdownCallbackType callback)
 {
-  on_shutdown_callbacks_.push_back(callback);
-  return callback;
+  std::lock_guard<std::mutex> lock(on_shutdown_callbacks_mutex_);
+  auto handle = std::make_shared<OnShutdownCallbackHandle>();
+  handle->callback = callback;
+  on_shutdown_callbacks_.push_back(handle);
+  return handle;
+}
+
+void
+Context::remove_on_shutdown_callback(OnShutdownCallbackHandle::SharedPtr callback_handle)
+{
+  std::lock_guard<std::mutex> lock(on_shutdown_callbacks_mutex_);
+  auto it = std::find_if(
+    on_shutdown_callbacks_.begin(),
+    on_shutdown_callbacks_.end(),
+    [callback_handle](const auto & weak_handle) {
+      return callback_handle.get() == weak_handle.lock().get();
+    });
+  if (it != on_shutdown_callbacks_.end()) {
+    on_shutdown_callbacks_.erase(it);
+  }
```

Comment:
I have to say that I don't love this; if there are a lot of handles in the list, this becomes a bottleneck.  But I guess until we prove that this is a performance problem, we can leave it as-is.

Reply:
:+1:
It doesn't matter much though, as I don't think that removing a shutdown callback is performance critical.

The underlying container could be a `std::unordered_set`, or something like that.

Reply:
I think that one reason to remove the `get_on_shutdown_callbacks()` methods is to not expose the underlying container type.

Reply:
Yeah. I follow the code style   
https://github.com/ros2/rclcpp/blob/b8df9347a10bff383fdc8890c5b1bfdf16430d8f/rclcpp/src/rclcpp/parameter_event_handler.cpp#L40-L56

But the performance is a problem. I will change to use `std::unordered_set`.